### ISSUE TITLE
Multi-cluster replicated control plane without gateway-connected clusters/ServiceEntry

### ISSUE THREAD

**Issue by bplotnick** [istio/istio#22605]
**Describe the feature request**
_I think this could probably be called "federated control plane" or something, but "federation" is so overloaded that we should probably come up with a better term._

**tl;dr:** I'm looking for a way to do multiple control planes but still have the same abstractions as in shared control plane - i.e. no `ServiceEntry`, no `.global` DNS pointing to an ingress gateway, etc.

I really like the [Shared Control Plane](https://istio.io/docs/setup/install/multicluster/shared/) model. Being able to stitch together clusters so that from the application networking perspective, cluster boundaries don't exist. However, this has the disadvantage of shared fate with cluster outages and/or having a single point of failure at the control plane.

To handle the resiliency, the recommended approach is to run multiple control planes. At the very least, people are running one-per-region. In some other cases, people are running one-per-az or one-per-cluster.

Unfortunately in the [Replicated Control Plane](https://istio.io/docs/setup/install/multicluster/gateways/) model, clusters are connected via gateways - meaning that the abstractions that are set up in the Shared Control Plane (`Service`, `VirtualService`, `Endpoint`) are not the same as those in the replicated control plane (`ServiceEntry`). Additionally, in a flat network, the ingress gateways for service-to-service communication are extraneous.

What I am proposing is having replicated control planes, but with "cross registration" of cluster resources (`Services`, `Endpoints`, whatever else Istio does within a cluster/control plane domain currently) from one cluster to another.

During a cluster failure or network partition, the only thing that will happen is that cross registration will stop, but since each control plane operates independently, the remaining control planes will still work fine.

You may now be asking about scaling concerns, but let's ignore it for now. The scaling problems will likely be the same as that of a single large control plane.

**Describe alternatives you've considered**
Admiral: It doesn't handle this I believe since it still deals with `ServiceEntry`s

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[X] Performance and Scalability
[X] Policies and Telemetry - Gateways hide the true endpoint that is being hit and so make the system less observable
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Comment by linsun** [comment]
@bplotnick thanks for raising this, we havesimilar thoughts as well - as you can tell from the discussion on the networking channel.  I think I agree with you with the notion of without service entries, I do think we need gateway connected clusters with auto passthrough for multi-network scenario, similar as what is in shared control plane multi network today.

I think we could possibly defer this (cross registration of istio resources) to a second phase as this can be driven from CI/CD system.

The first phase issue is how to configure Endpoint and services across clusters when each cluster has its own control plane.  @costinm mentioned by connecting to all apiservers, which I think it is doable.  let us say i have c1 in region east, and c2 in region south.   I can give c1 to access c2's API server via istio-reader-service-account as described in https://istio.io/docs/setup/install/multicluster/shared/#configure-cross-cluster-service-registries.  Similarly, I can give c2 access to c1's API server.  @costinm could you confirm that is what you are thinking?   

cc @ayj @howardjohn @frankbu @esnible @mbanikazemi @dcberg

**Comment by howardjohn** [comment]
I recall seeing some discussion about how this approach is not preferred due to the n^2 connections from all control planes to all api servers? From @nmittler I believe

**Comment by linsun** [comment]
Agree if you have many clusters and many API servers this can be lots of connections.  I don't think this would be used often, mainly across regions.  within region, you can use single control plane.  The prob we have is we don't want to rely on single control plane across regions.

**Comment by ayj** [comment]
In principle there are several dimensions to multi-cluster deployment models:

* shared vs. replicated (where do control planes run)
* service discovery (explicit ServiceEntry vs. directly peer kube-apiservers)
* number of networks

Not every combination is supported, primarily we don't test every combination. Here's what we document on istio.io:

1. Shared, single network, service/endpoints directly from peer kube-apiservers
2. Shared, multiple network, service/endpoints directly from peer kube-apiservers
3. Replicated, multiple network, service/endpoints explicitly from ServiceEntry

IIUC, @bplotnick is asking about the following model:

4. Replicated, single network, service/endpoints directly from peer kube-apiservers

This is similar to (1) and in fact easier to configure than (1). k8s services/namespaces still need to be replicated so DNS is configured correctly. 

```
# <... install intermediate certs with common root in each cluster ...>
istioctl manifest apply --context=cluster0 --set values.global.multiCluster.clusterName=cluster0
istioctl manifest apply --context=cluster1 --set values.global.multiCluster.clusterName=cluster0
istioctl x create-remote-secret --context=cluster0 --name=cluster0 | kubectl apply -f - --context=cluster1
istioctl x create-remote-secret --context=cluster1 --name=cluster1 | kubectl apply -f - --context=cluster0
```

Also worth noting that multi-network variant of replicated control plane using kube-apiserver is also possible. 

5. Replicated, multiple network, service/endpoints directly from peer kube-apiservers

This requires more steps to configure the networks. The basic idea is to follow  https://istio.io/docs/setup/install/multicluster/shared/#main-cluster to create _both_ cluster's profile. 

For example, cluster0's network config may look like:

```
    global:
      multiCluster:
        clusterName: cluster0
      network: network0
      meshNetworks:
        network0
          endpoints:
          # registry name for the cluster that this configuration is applied to must be "Kubernetes"
          - fromRegistry: Kubernetes
          gateways:
          - registry_service_name: istio-ingressgateway.istio-system.svc.cluster.local
            port: 443
        network1
          endpoints:
          # registry name must match cluster1's global.multiCluster.clusterName
          # and remote secret name
          - fromRegistry: cluster1       
          gateways:
          - registry_service_name: istio-ingressgateway.istio-system.svc.cluster.local
            port: 443
```

And cluster1's network config may look:

```
    global:
      multiCluster:
        clusterName: cluster1
      network: network1
      meshNetworks:
        network0
          endpoints:
          # registry name must match cluster0's global.multiCluster.clusterName
          # and remote secret name
          - fromRegistry: cluster0
          gateways:
          - registry_service_name: istio-ingressgateway.istio-system.svc.cluster.local
            port: 443
        network1
          endpoints:
          # registry name for the cluster that this configuration is applied to must be "Kubernetes"
          - fromRegistry: Kubernetes  
          gateways:
          - registry_service_name: istio-ingressgateway.istio-system.svc.cluster.local
            port: 443
```

Then, apply the custom profiles to each cluster and create the remote secrets.

# <... install intermediate certs with common root in each cluster ...>
istioctl manifest apply --context=cluster0 -f <cluster0-profile.yaml>
istioctl manifest apply --context=cluster1 -f <cluster1-profile.yaml>
istioctl x create-remote-secret --context=cluster0 --name=cluster0 | kubectl apply -f - --context=cluster1
istioctl x create-remote-secret --context=cluster1 --name=cluster1 | kubectl apply -f - --context=cluster0
```

**Comment by ayj** [comment]
Please don't copy/paste the config/commands above :) There's likely a typo (or two) and I may have forgot some minor step. But the basic flow is still sound.

As @howardjohn mentioned, one downside to this approach is Ω(N^2) watches between clusters. The number is higher when there are multiple istiod pods per-cluster. It also introduces yet another deployment model when we're actively trying to simply multi-cluster configuration. That said, it's may still be worth exploring how to make this work without overwhelming users with options.

cc @nmittler @smawson @rcaballeromx

**Comment by linsun** [comment]
@ayj thanks for sharing the steps, that is pretty much I have in mind as well.  It is a mixed model with our single control plane and replicated control plane model.  Certainly I am interested in this as I don't have flat network env:

- multiple control planes, multiple networks, service/endpoints directly from peer kube-apiservers

In this model, users don't need to worry about creating SEs or VSs.  

Re N^2 watches between clusters, one way to tackle this is to not auto expose everything across clusters, could be only a given global namespace or only services with certain matching label e.g. gobal label.

**Comment by bplotnick** [comment]
Thanks for the discussion. And special thanks to @costinm for the discussion in Slack.

> I think we could possibly defer this (cross registration of istio resources) to a second phase as this can be driven from CI/CD system.

After the discussion in slack, I agree that this is out of scope. Endpoint discovery[0] is the main thing we need. Just like `ServiceEntry` today, we can leave replicating namespaces, `Service`, `VirtualService` resources, etc. to the users for now.

> I recall seeing some discussion about how this approach is not preferred due to the n^2 connections from all control planes to all api servers?

I think in terms of scaling, there are ways that we can bring these factors down (e.g. designated aggregators, which a networking colleague of mine likened to [BGP Route Reflectors](https://en.wikipedia.org/wiki/Border_Gateway_Protocol#Route_reflectors)). But for now, let's acknowledge the potential scaling issues exist and move on.

> It also introduces yet another deployment model when we're actively trying to simply multi-cluster configuration

I agree. The last thing Istio needs is more complexity. But I actually think this has the potential to replace the `ServiceEntry` approach and reduce complexity. In fact the reduction in complexity is one of the main reasons that I like this approach. Given that a `ServiceEntry` is for external services, it was very strange to me to use it for internal services. And to have a separate `.global` name. In my ideal world, something like [Admiral](https://github.com/istio-ecosystem/admiral) doesn't need to exist, or would be focused on something else.

If we could annotate endpoints with the cluster/network then we could use a network topology to determine routes to endpoints - i.e. whether to go through a passthrough gateway. Though to be honest, this is irrelevant to my use-case as we have a flat network, so i don't know what this type of user wants.

[0] Terminology note (for noobs like me): Coming from the SmartStack world, we distinguished "Service Registration" from "Service Discovery". It appears that this distinction is not as relevant in the Kubernetes world because the registration is automatically done by K8s. Additionally, because `Service` is a separate thing from `Endpoint`, the term should probably be something more like "Endpoint Discovery".

**Comment by ayj** [comment]
> If we could annotate endpoints with the cluster/network then we could use a network topology to determine routes to endpoints - i.e. whether to go through a passthrough gateway.

FWIW, this is how shared control plane works today w/single and multi-network.

> But for now, let's acknowledge the potential scaling issues exist and move on.

For smaller values of N (clusters) the extra control plane redundancy is probably worth the cost of extra connections.

>  But I actually think this has the potential to replace the ServiceEntry approach and reduce complexity. In fact the reduction in complexity is one of the main reasons that I like this approach. 

I would be interested to hear others thoughts here. It is appealing to reduce the dimensions to "shared vs. replicated" and "# of networks".

**Comment by linsun** [comment]
I think I'm interested in this approach too and agree it is a simplification, managing service entries aren't easy when there is more than 2 clusters.... plus it doesn't support weighted based routing for a given subset because you don't have visibility into the other cluster like the current shared control plane does.

If we can introduce a reasonable multiple control plane single or multiple network approach, i think it could possibly replace the replicated control plane and we will converge on one model and let users to decide if they choose to run shared control plane vs multiple control planes.

In summary, it would be 1 single model, while up to users to choose:
1. shared control plane vs dedicated control plane
2. single or multi networks
3. choose what to expose to the other control planes so not necessarily everything is exposed... Kube community already has an API in progress: https://docs.google.com/document/d/1hFtp8X7dzVS-JbfA5xuPvI_DNISctEbJSorFnY-nz6o/edit#heading=h.u7jfy9wqpd2b 

(A side note, I think replicated control plane naming is pretty confusing... it is just single trust domain and multiple control planes imho.)

**Comment by sdake** [comment]
> Please don't directly copy/paste the config/commands above :) There's likely a typo (or two) and I probably forgot some minor step.
> 
> As @howardjohn mentioned, one downside to this approach is Ω(N^2) watches between clusters. The number is higher when there are multiple istiod pods per-cluster. It also introduces yet another deployment model when we're actively trying to simply multi-cluster configuration. That said, it may still be worth exploring how to make this all work without overwhelming users with options.
> 
> cc @nmittler @smawson @rcaballeromx

Catching up on this issue which contains many subthreads...

The fascinating beauty of using a ServiceEntry to represent an external workload in an opposing mesh is that O(N^2) watches are not required on the kube-apiservers to enable service discovery.

Instead each cluster is responsible for watching its own kube-apiserver. Each cluster reads service creation/removal and communicates the results from kube-apiserver (which is a pokey rate-limited REST API with complex caching mechanics) to its opposing cluster using a high-performance well-designed gRPC between two clusters. The opposing cluster then creates service entries on the opposing end representing the services. If these are connected in opposing fashion, as all good networking is done, any network topology can be modeled.

Using this model scales, even though there are O(N^2) gRPC connections. If we can read from kube-apiserver one time for each service change, with the reading happening within each cluster, and then encode that information into gRPC transmissions, The O(N^2) that was kubeapi turns into O(1) in kubeapi and O(N^2) into gRPC calls in the worst case (where N = number of clusters). As any Kubernetes expert knows, kubeapi is what we optimize away.

Plus, we don't have to security nightmare of having 100's of kube-apiserver credentials stored across hundreds of hypothetical clusters.

Put simply, reading from kube-apiserver for every service creation and deletion is excessively expensive, and we need to offload that.  About a year ago, this was what was to be the purpose of MCP. That never materialized, and I think it is about time to solve replicated control planes in the community.

This is the most scalable path forward.
This model also offers the most availability.

**Comment by sdake** [comment]
> @ayj small q for what you mentioned = `"k8s services/namespaces still need to be replicated so that DNS works correctly ." This is only required if i need cross cluster routing correct?

The RFC1918 networking model (what has been merged - more or less) replication problem is about discovery. For folks that haven't been following this area closely, discovery is the major problem to be solved.

The model @rshriram proposed and implemented was really really close. Where it could be improved:

1. service entries were not automatically managed
2. cloud providers with loads of extra DNS entries would run into application discovery problems.

proposal https://github.com/istio/istio/issues/22605#issuecomment-608950278 solves 1.
PR: https://github.com/istio/istio/pull/22369#issuecomment-604513008 (see the point about how global is handled in istiod instead of being added as a search domain!) solves 2.

I'll be honest when I say that I have about thrown in on the gateways approach to multicluster. @costinm, fortunately, saved the day (again) with his PR proposal. Doing 1 is fairly easy. Problem 2 seemed unsolvable.

I feel we should keep them separate a bit while longer until we can properly determine the best course of action once Costin's DNS work merges.

Cheers
-steve

**Comment by frankbu** [comment]
There is no O(N^2) scalability issue if we don't look at replicated control plane as being an all or nothing choice (i.e., shared vs replicated). Replicated control planes are needed for HA, but that doesn't mean that we need 1 in every cluster, just more than 1. In this case, it's only O(C*N) connections, which can be much less than N^2. So I would think we should be focusing on a replicated pattern as an HA addition to the shared pattern, i.e., one where a subset, as opposed to just 1, of the clusters run a control plane. That said, @bplotnick's proposal to eliminate ServiceEntry from the pattern seems like a great simplification.

**Comment by dcberg** [comment]
Catching up on this thread. I do agree that the important point is to support Endpoint discovery and avoid having customers define ServiceEntries to use remote services. I do agree with @linsun comment (https://github.com/istio/istio/issues/22605#issuecomment-608629319) that the customer  can have shared vs dedicated control planes and single vs multiple networks.

What is the next step here?

**Comment by nmittler** [comment]
Just catching up as well. IIUC the root cause of this issue is that the istio.io docs conflate endpoint discovery with control plane deployments. In other words, it appears that the only way to deploy replicated control planes is with gateways + ServiceEntry. 

We have already documented various dimensions of Istio deployments in the [deployment models concepts page](https://istio.io/docs/ops/deployment/deployment-models/), where we identify things such as network and control plane topology. We don't currently, however, mention Endpoint Discovery as a dimension. This is probably due to the fact that it's more of a derived dimension, based on more fundamental dimensions (e.g. control plane, cluster, network).

I wonder if we could resolve this issue by simply updating the docs. We could do the following:

- Add a section to the concepts page for Endpoint Discovery
- Update installation instructions (for both shared and replicated control planes) to configure Endpoint Discovery based one one of two approaches (depending on network/control plane configuration).

Obviously we still have more to do WRT simplifying the multicluster installation overall, but this should at least help users avoid needless complexity in the meantime.

WDYT?

**Comment by dcberg** [comment]
@nmittler I believe that would be a good first start (i.e., document how to achieve Endpoint discovery with the existing models).

**Comment by dcberg** [comment]
I have a related question. If you start with a replicated deployment (2 clusters with separate meshes) can you then use one of the meshes as a shared control plane for a third cluster?

**Comment by nmittler** [comment]
> I have a related question. If you start with a replicated deployment (2 clusters with separate meshes) can you then use one of the meshes as a shared control plane for a third cluster?

To be clear, replicated control planes don't imply "separate meshes". A separate mesh typically implies a lot of other things (e.g. separate trust domain). In general, when we talk about shared vs replicated control planes, we mean that it's all a single mesh.

But to the question regarding adding a new cluster as "shared" ... Yes, absolutely. I had discussed an approach a while back with @rcaballeromx about the possibility of changing the way we document installations so that they are more "legos". The idea is:

1. Start with a single cluster with a control plane
2. Then, add a cluster to transition to another state. Here we have options for same/different network, or have a shared/replicated control plane.
3. From that point, adding yet another cluster is just rinse and repeat (for the most part).

This may be an option to help simplify our installation docs a bit.

**Comment by dcberg** [comment]
@nmittler you are correct. I should have been more clear there I start with one logical service mesh with replicated control planes (shared trust).
Having the documentation where you can assemble your service mesh like legos over time would be nice. Today the documentation reads as if you are always starting from scratch and building exactly what you need from day one which is unrealistic for most users.

**Comment by linsun** [comment]
An initial validation through of setting up 2 clusters with multi-network and each cluster has its own istiod for xDS serving, sidecar injector and CA, it looks positive and working on IKS.  No need to create service entries in this model and endpoints are auto discovered across 2 clusters

I haven't got locality lb working which I think it is due to my clusters are both in the same zone.  I will spin another cluster in a different zone and try that too so we can ensure we use local service first and fail over to cluster in different zone/region when local isn't avail.

**Comment by linsun** [comment]
@ayj thank you for the steps, they were very helpful

**Comment by ramaraochavali** [comment]
@linsun Just curious, how do you access the service in remote cluster - more interested in how DNS works for `cluster.local` vs `.global` without service entry?

**Comment by sdake** [comment]
@ramaraochavali this is an area for development. Prototypes from the 1.1 branch show routing crashes when route discovery takes place with a replicated service in the replicated control plane collide. Please see this issue for more context: https://github.com/istio/istio/issues/13045. I suspect this is what @linsun was speaking about.

Cheers
-steve

**Comment by linsun** [comment]
@ramaraochavali @sdake If you are using multiple clusters for service failover or cross cluster routing scenarios, you will most likely have the same service in the same namespace in cluster 1 and cluster 2.  That would create the DNS entries and istio can help you with endpoint discoveries by watching kube API servers from remote clusters similarly as what is described in the "shared control plane multicluster model" except each cluster will have its own control plane.

**Comment by ramaraochavali** [comment]
@sdake Thank you for the details. Will take a look

@linsun The use case I was thinking is cross cluster routing i.e. foo in cluster1 -> bar in cluster 2. So IIUC, you are suggesting that foo,bar both should be in cluster1, cluster2. Is that correct?

**Comment by frankbu** [comment]
This issue is addressed in 1.8: https://preliminary.istio.io/latest/docs/setup/install/multicluster/

Also related: https://preliminary.istio.io/latest/blog/2020/dns-proxy/
